neqo-transport/src/connection.rs

```diff
@@ -415,26 +417,46 @@ impl Connection {
 
     /// Call in to process activity on the connection. Either new packets have
     /// arrived or a timeout has expired (or both).
-    pub fn process_input(&mut self, dgram: Datagram, now: Instant) {
-        let res = self.input(dgram, now);
-        self.absorb_error(now, res);
+    pub fn process_input(&mut self, dgram: Option<Datagram>, now: Instant) {
+        if let Some(idle_timeout) = self.idle_timeout.take() {
+            if now >= idle_timeout {
+                self.set_state(State::Closing {
+                    error: ConnectionError::Transport(Error::IdleTimeout),
+                    frame_type: 0,
+                    msg: "Idle timeout".into(),
+                    timeout: self.get_closing_period_time(now),
+                });
+            }
+        }
+
+        if let Some(dgram) = dgram {
+            let res = self.input(dgram, now);
+            self.absorb_error(now, res);
+        }
         self.cleanup_streams();
         self.check_loss_detection_timeout(now);
     }
 
     /// Get the time that we next need to be called back, relative to `now`.
-    fn next_delay(&mut self, now: Instant) -> Option<Duration> {
+    fn next_delay(&mut self, now: Instant) -> Duration {
         self.loss_recovery_state = self.loss_recovery.get_timer(&self.state);
+        let ack_time = self.acks.ack_time();
 
-        let time = match (self.loss_recovery_state.callback_time, self.acks.ack_time()) {
-            (Some(t_lr), Some(t_ack)) => Some(min(t_lr, t_ack)),
-            (Some(t), _) | (_, Some(t)) => Some(t),
-            _ => None,
+        qdebug!([self] "LR timer?: {:?} ack timer?: {:?}", self.loss_recovery_state.callback_time, ack_time);
+
+        let time = match (self.loss_recovery_state.callback_time, ack_time) {
+            (Some(t_lr), Some(t_ack)) => min(t_lr, t_ack),
+            (Some(t), _) | (_, Some(t)) => t,
+            (None, None) => {
+                let timeout = now + Duration::from_millis(LOCAL_IDLE_TIMEOUT);
```

Comment:
LOCAL_IDLE_TIMEOUT could be a Duration, right?  Then you don't need to worry about getting units right.